pkg/resources/grant_privileges_to_database_role.go

```diff
@@ -0,0 +1,1031 @@
+		ExactlyOneOf: []string{
+			"on_database",
+			"on_schema",
+			"on_schema_object",
+		},
+	},
+	"on_schema": {
+		Type:        schema.TypeList,
+		Optional:    true,
+		ForceNew:    true,
+		Description: "Specifies the schema on which privileges will be granted.",
+		MaxItems:    1,
+		ExactlyOneOf: []string{
+			"on_database",
+			"on_schema",
+			"on_schema_object",
+		},
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"schema_name": {
+					Type:             schema.TypeString,
+					Optional:         true,
+					ForceNew:         true,
+					Description:      "The fully qualified name of the schema.",
+					ValidateDiagFunc: IsValidIdentifier[sdk.DatabaseObjectIdentifier](),
+					ExactlyOneOf: []string{
+						"on_schema.0.schema_name",
+						"on_schema.0.all_schemas_in_database",
+						"on_schema.0.future_schemas_in_database",
+					},
+				},
+				"all_schemas_in_database": {
+					Type:             schema.TypeString,
+					Optional:         true,
+					ForceNew:         true,
+					Description:      "The fully qualified name of the database.",
+					ValidateDiagFunc: IsValidIdentifier[sdk.AccountObjectIdentifier](),
+					ExactlyOneOf: []string{
+						"on_schema.0.schema_name",
+						"on_schema.0.all_schemas_in_database",
+						"on_schema.0.future_schemas_in_database",
+					},
+				},
+				"future_schemas_in_database": {
+					Type:             schema.TypeString,
+					Optional:         true,
+					ForceNew:         true,
+					Description:      "The fully qualified name of the database.",
+					ValidateDiagFunc: IsValidIdentifier[sdk.AccountObjectIdentifier](),
+					ExactlyOneOf: []string{
+						"on_schema.0.schema_name",
+						"on_schema.0.all_schemas_in_database",
+						"on_schema.0.future_schemas_in_database",
+					},
+				},
+			},
+		},
+	},
+	"on_schema_object": {
+		Type:        schema.TypeList,
+		Optional:    true,
+		ForceNew:    true,
+		Description: "Specifies the schema object on which privileges will be granted.",
+		MaxItems:    1,
+		ExactlyOneOf: []string{
+			"on_database",
+			"on_schema",
+			"on_schema_object",
+		},
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"object_type": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					ForceNew:    true,
+					Description: "The object type of the schema object on which privileges will be granted. Valid values are: ALERT | DYNAMIC TABLE | EVENT TABLE | FILE FORMAT | FUNCTION | PROCEDURE | SECRET | SEQUENCE | PIPE | MASKING POLICY | PASSWORD POLICY | ROW ACCESS POLICY | SESSION POLICY | TAG | STAGE | STREAM | TABLE | EXTERNAL TABLE | TASK | VIEW | MATERIALIZED VIEW | NETWORK RULE | PACKAGES POLICY | ICEBERG TABLE",
+					RequiredWith: []string{
+						"on_schema_object.0.object_name",
+					},
+					ConflictsWith: []string{
+						"on_schema_object.0.all",
+						"on_schema_object.0.future",
+					},
+					ValidateFunc: validation.StringInSlice([]string{
```

Comment:
Hmm, some of the fields may require different sets of object types, but it could be also used in the snowflake_grant_privilges_to_role. I'll extract it and use it there.